### PR TITLE
fix: camera remains active after closing video recording modal

### PR DESCRIPTION
**Pull request by abhinavkrin**
## Proposed changes (including videos or screenshots)

When users close the video recording modal quickly before the camera has fully initialized, the camera light remains on until the page is refreshed. This occurs due to a race condition where stop() is called before the asynchronous getUserMedia() callback completes, causing the stream to initialize after cleanup has been attempted

- Implemented a session ID pattern to track and invalidate stale async callbacks:
- Added a sessionId counter that increments on each start() and stop() call
- The getUserMedia success callback validates its session ID before initializing the stream
- If the session ID has changed (indicating stop() was called), the stream is immediately cleaned up

##### Why Session ID Mechanism is Needed
Without a way to invalidate pending async callbacks, the original code had no mechanism to prevent stale getUserMedia success handlers from executing after stop() was called. Even though stop() cleared the stream reference, the pending callback would still execute and reinitialize the camera, leaving it active indefinitely.

https://github.com/user-attachments/assets/c7d21d07-b849-405c-a160-62a9d237fc3a

## Further comments

[CORE-1301](https://rocketchat.atlassian.net/browse/CORE-1301)

[CORE-1301]: https://rocketchat.atlassian.net/browse/CORE-1301?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

* **Bug Fixes**
  * Camera now reliably stops when closing the video recording modal; prevents the camera remaining active.
* **Tests**
  * Added comprehensive unit tests for the video recorder covering start/stop cycles, race conditions, permission errors, and stream handling.
* **Refactor**
  * Improved start/stop logic to avoid lingering streams and invalidate stale operations.
* **Chores**
  * Expanded test discovery to include additional client-side spec files.

### CODE REVIEW

File: apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+		global.MediaRecorder = {
+			isTypeSupported: jest.fn((type: string) => type === 'video/webm'),
+		} as any;
+	});
+
+	afterEach(() => {
+		jest.clearAllMocks();
+	});
+
+	describe('Race condition fix', () => {
```
Comment (aleksandernsilva) [inline apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts]: Although this is a test suite aimed at the race condition fix, describing it like that seems very specific imo. I recommend changing to something along the lines of `start and stop edge case handling` or `Asynchronous start and stop handling`, etc.

Something that describes what we are testing instead of why the tests were added.

File: apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+		it('should stop camera tracks when stop is called before getUserMedia resolves', async () => {
+			const streamPromise = new Promise<MediaStream>((resolve) => {
+				setTimeout(() => resolve(mockStream), 100);
+			});
```
Comment (aleksandernsilva) [inline apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts]: You could reduce the use of `setTimeout` in this file by using the util [createDeferredPromise](https://github.com/RocketChat/Rocket.Chat/blob/develop/apps/meteor/tests/mocks/utils/createDeferredMockFn.ts). In this this instance the implementation could change to something like:
```ts
const streamDeferred = createDeferredPromise<MediaStream>();

getUserMediaMock.mockReturnValue(streamDeferred.promise);

// [...]

streamDeferred.resolve(mockStream);

// [...]
```

File: apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+			getUserMediaMock.mockReturnValue(streamPromise);
+
+			const callback = jest.fn();
+			VideoRecorder.start(mockVideoElement, callback);
+			VideoRecorder.stop();
+
+			await streamPromise;
+			await new Promise((resolve) => setTimeout(resolve, 150));
```
Comment (aleksandernsilva) [inline apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts]: Could we take advantage of [jest fake timers](https://jestjs.io/docs/jest-object#fake-timers) and [jest.runAllTicks](https://jestjs.io/docs/jest-object#jestrunallticks) to eliminate some of these arbitrary timeouts?

File: apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+		it('should handle multiple start/stop cycles', async () => {
+			const stream1 = {
+				getVideoTracks: jest.fn(() => [{ stop: jest.fn() } as unknown as MediaStreamTrack]),
+				getAudioTracks: jest.fn(() => [{ stop: jest.fn() } as unknown as MediaStreamTrack]),
+			} as unknown as MediaStream;
+
+			const stream2 = {
+				getVideoTracks: jest.fn(() => [mockVideoTrack]),
+				getAudioTracks: jest.fn(() => [mockAudioTrack]),
+			} as unknown as MediaStream;
```
Comment (aleksandernsilva) [inline apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts]: Nitpick: You could create a small utility function to facilitate creating these mock streams. Would improve code readability as well to some extent,

File: apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+		mockVideoTrack = {
+			stop: jest.fn(),
+		} as unknown as MediaStreamTrack;
+
+		mockAudioTrack = {
+			stop: jest.fn(),
+		} as unknown as MediaStreamTrack;
+
+		mockStream = {
```
Comment (aleksandernsilva) [inline apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts]: mockStream = createMockStream(...)

File: apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+			getUserMediaMock.mockReturnValue(streamDeferred.promise);
+
+			const callback = jest.fn();
+			VideoRecorder.start(mockVideoElement, callback);
+			VideoRecorder.stop();
+
+			streamDeferred.resolve(mockStream);
+			await Promise.resolve();
+			jest.runAllTimers();
+			await Promise.resolve();
```
Comment (aleksandernsilva) [inline apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts]: `await Promise.resolve()` is doing all the heavy lifting here, rendering the fake timers unnecessary. This applies to all other tests in this file.
```suggestion
			await jest.runAllTimersAsync();
```

File: apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+		mockAudioTrack = {
+			stop: jest.fn(),
+		} as unknown as MediaStreamTrack;
+
+		mockStream = {
+			getVideoTracks: jest.fn(() => [mockVideoTrack]),
+			getAudioTracks: jest.fn(() => [mockAudioTrack]),
+		} as unknown as MediaStream;
+
+		mockVideoElement = document.createElement('video');
```
Comment (aleksandernsilva) [inline apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts:43]: jsdom doesn't seem to support any loading or playback media operations. When running these tests it ends up generating a lot of `console.error`'s which makes it difficult to read the logs. Could you mock the necessary functions as well?

File: apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+			getUserMediaMock.mockReturnValueOnce(firstDeferred.promise).mockReturnValueOnce(secondDeferred.promise);
+
+			const cb1 = jest.fn();
+			const cb2 = jest.fn();
+
+			VideoRecorder.start(mockVideoElement, cb1);
+			VideoRecorder.start(mockVideoElement, cb2);
+
+			secondDeferred.resolve(secondStream);
+			await Promise.resolve();
```
Comment (aleksandernsilva) [inline apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.spec.ts]:
```suggestion
			await jest.runAllTimersAsync();
```

**Review by aleksandernsilva — CHANGES_REQUESTED** [review]

**Review by aleksandernsilva — APPROVED** [review]

File: apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.ts
Comment (dougfabris) [inline apps/meteor/app/ui/client/lib/recorderjs/videoRecorder.ts:1]: I liked the approach here, but I have to say your code doesn't follow the principle of few comments, here there are some considerations from Robert C. Martin, please consider following them:

**Comments are Failures**: a comment indicates you couldn't make the code express its purpose. Instead of writing a comment, aim to refactor the code to be clearer.

**Comments Lie**: Comments rarely get maintained, so they quickly become inaccurate, misleading, or outdated as the code evolves.

**Avoid Redundant/Bad Comments**: Avoid comments that just explain what the code does (it should be obvious), journal comments (version history), or commenting out code.

**When They Are Okay**: They can be used for "essential" information that cannot be expressed in code, such as legal notices, providing intent, or explaining a particularly complex, un-refactorable algorithm.

**TODO Comments**: He advises using TODO comments only for temporary notes, not leaving them in the repository.

**Code Should Speak**: Good code should be self-explanatory. Prefers clear variable/function names over comments. 

Additionally, I would suggest reading [this](https://blog.cleancoder.com/uncle-bob/2017/02/23/NecessaryComments.html) post
